EchoPro/core.py

```diff
@@ -0,0 +1,174 @@
+"""
+Core API/components for class structures, class methods, and utility/helper methods
+"""
+import numpy as np
+
+CONFIG_MAP = {
+    """
+    ``CONFIG_MAP`` defines the expected column names for all datasets defined within both
+    ``initialization_config.yml`` and ``survey_year_2019.yml``. This further pairs the
+    expected datatype with each variable. The nested dictionary names represent 
+    tags that are queried when the data are loaded which reflect the naming conventions
+    found within the configuration files.
+    """
+    # TODO: Update documentation to reflect dynamic parameterization since not all datasets
+    # will be relegated to just the year 2019 upon deployment.
+
+    # BIOLOGICAL DATASET -- LENGTH
+    "biological.length": {
```

Comment:
The use of `.` is pretty common in R, but I think it would be good to stick with "_" for variables in Python.
```suggestion
    "biological_length": {
```

Reply:
Perhaps this may make more sense in the context of how I implemented `CONFIG_MAP` for `Survey` parameterization, but these dictionary key names with `CONFIG_MAP` are primarily used as tags to help link `CONFIG_MAP` to a specific data file for validation (e.g. length data for US fish). So in this case, text on each side of the `.` delimitor correspond to first two nested layers as defined in `survey_year_2019_config.yml`. So for instance: `'biological.length.US'` maps to the configuration settings that yield: `{'filename': 'Biological/US/2019_biodata_length.xlsx', 'sheetname': 'biodata_length'}`. Similarly, `biological.length` links the values within `CONFIG_MAP` to all files associated with `biological.length`, so in this case: `biological.length.US` and `biological.length.CAN`. I elected to use `.` over `_` since some nested dictionary key names also contain `_`, such as `all_ages'`. So `NASC.all_ages`, as an example, was easier to parse/separate than `NASC_all_ages` (at least with how the code is currently written) when needing to toggle between `NASC.all_ages` and `[NASC, all_ages]`. Since these tags are really representing the dictionary paths of the stored dataframes, perhaps it would make more sense to replace `.` with `/`? For reference, I have included a rough schematic of the data attribute structure:

* 'acoustics': dict
    * 'nasc': dict
        * 'no_age1_df': DataFrame, 
        * 'all_ages_df': DataFrame,
* 'biology': dict 
    * 'catch_df': DataFrame
        * ['CAN', 'US']: regions within df
    * 'haul_to_transect_df': DataFrame
        * ['CAN', 'US']: regions within df
    * 'length_df': DataFrame 
        * ['CAN', 'US']: regions within df
    * 'specimen_df': DataFrame 
        * ['CAN', 'US']: regions within df
* 'spatial': dict
    * 'strata_df': DataFrame, 
    * 'geo_strata_df': DataFrame,
* 'statistics': dict
    * 'kriging': dict 
        - 'mesh_df': DataFrame,
        - 'isobath_200m_df': DataFrame,
        - 'vario_krig_para_df': DataFrame

So if I wanted to get the NASC exports for all age-classes, the nested dictionary path would be represented by either: `'acoustics/nasc/all_ages_df'`,`'acoustics.nasc.all_ages_df'`, or `'acoustics_nasc_all_ages_df'`. Similarly, `'biology.length_df'` in the above structure would get the concatenated DataFrame containing both the `US` and `CAN` datasets. I have no problem with using `_` instead; however, it would require some tweaking to either the `survey_year_2019_config.yml` or the code that define, initialize, and parameterize `Survey` in `survey.py`. With all of this in mind, I personally think `/` may be a better delimiter in this case since it is a bit more intuitive to associate `/` with file paths (or in this case the dictionary paths) versus `.` or `_`.

Reply:
I see, I can see the problem with having everything with `_`. 

If your goal is to map to what's in `survey_year_2019_config.yml`, how about using the same nested dictionary structures as what would be read in using `yaml.load` (currently used under `_read_config)`:
```python
CONFIG_MAP = {
    "biological": {
        # BIOLOGICAL DATASET -- LENGTH
        "length" : {
            "haul_num": int ,
            "species_id": int ,
            "sex": np.uint8 ,
            "length": np.float64 ,
            "length_count": int ,
        },
        # BIOLOGICAL DATASET -- CATCH
        "catch": {
            "haul_num":  int ,
            "species_id": int ,
            "haul_weight": np.float64 ,
        } ,
    },
    "stratification": {
        "strate" : {
            ...
        },
        "geo_strata": {
            ...
        }
    },
    ...
}
```
So that you could use exactly the same string to index both.

One of the recent changes Emilio did (in https://github.com/uw-echospace/EchoPro/pull/121) was exactly to enable splitting strings with `/` so that the `survey_year_2019_config.yaml` inputs is more succinct. So it could probably be that the underlying structure in `core` is the nested dictionaries, but when you write it in code, it would just be the strings with `/` delimiters with the same kind of splitting when it is used.